intro-to-object-oriented-programming/instance-methods.md

```diff
@@ -0,0 +1,409 @@
+| Album  | Dirty Computer (2018), which has 14 tracks in a certain order, was released April 2018, has a total album length of 48:42 | Return the audio data for any specific track it has                          |
+
+## Defining Instance Methods
+
+Because instance methods are functions inside of the class definition, instance methods can use `self`... In fact, in some ways it _must_!
+
+Instance methods are defined with similar syntax to other functions. But, just like the `__init__` method, the first parameter listed **must** be `self`.
+
+```python
+class ExampleClassName:
+
+    def example_instance_method(self):
+        pass
+```
+
+- We can specify a `return` value!
+- We can add as many parameters we want after `self`
+- Inside of this method, we can access attributes and _other_ methods, using `self`
+- We can do any other typical Python logic in here, such as loops, conditionals, etc.
+
+### !callout-info
+
+## `self` Must Be The First Parameter
+
+Just like the `__init__` method, the first parameter listed **must** be `self`.
+
+### !end-callout
+
+Observe these examples for an `Album` class. The examples increase in complexity.
+
+```python
+class Album:
+
+    def get_audio_data(self):
+        return "Prince - When Doves Cry (5:54) (Track 1) (Genre: Pop, rock, R&B, psychedelia)"
+```
+
+In the above example, every instance of `Album` has the instance method `get_audio_data`. Calling this method will always return `"Prince - When Doves Cry (5:54) (Track 1) (Genre: Pop, rock, R&B, psychedelia)"`.
+
+```python
+class Album:
+
+    def __init__(self, title, release_date, tracks):
+        self.title = title
+        self.release_date = release_date
+        self.track_list = tracks
+
+    def get_audio_data(self):
+        return self.track_list
+```
+
+In the above example, the instance method `get_audio_data` reads the attribute `track_list`. It returns and reads this attribute with `return self.track_list`. It will return whatever the value of `self.track_list` is, which was defined in the `__init__` method; it was assigned to `tracks`.
+
+```python
+class Album:
+
+    def __init__(self, title, release_date, tracks):
+        self.title = title
+        self.release_date = release_date
+        self.track_list = tracks
+
+    def get_audio_data(self, track_index):
+        if len(self.track_list) > 0:
+            return self.track_list[track_index - 1]
+        else:
+            return "The track list is empty"
+```
+
+In the above example, the instance method `get_audio_data` takes an additional parameter `track_index`. This means that whenever we call this method, we must pass in an argument for `track_list`. `track_list` will represent the index track we want audio data from.
+
+Inside the `get_audio_data` method, we use the attribute `self.track_list` _**and**_ the argument `track_index`. We assume `self.track_list` is a list, and we use a conditional statement.
```

Comment:
I'm open to the suggestion if you'd push for it...

In general, I like excluding it/opting for this "simpler" "solution" since the meat of this lesson is about instance methods, but realistically I like the code you're suggesting more

Reply:
What about this? Safe, and it gets to show a three-way boolean!
```python
def get_audio_data(self, track_index):
    if 0 < track_index <= len(self.track_list):
        return self.track_list[track_index - 1]
    else:
        return "Invalid track"
```